core/src/main/java/io/grpc/ClientCallImpl.java

```diff
@@ -204,14 +299,42 @@ private void closeCallPrematurely(ClientStreamListener listener, Status status)
   }
 
   private ScheduledFuture<?> startDeadlineTimer(long timeoutMicros) {
-    return deadlineCancellationExecutor.schedule(new Runnable() {
+    return scheduledExecutor.schedule(new Runnable() {
       @Override
       public void run() {
-        stream.cancel(Status.DEADLINE_EXCEEDED);
+        // races with the future returned by this method, but should be okay as long as the
+        // scheduler is single threaded.
+        if (retryFuture != null) {
+          retryFuture.cancel(true);
+        }
+        synchronized (ClientCallImpl.this) {
+          if (stream != null) {
+            stream.cancel(Status.DEADLINE_EXCEEDED);
+          } else {
+            callListener.onClose(Status.DEADLINE_EXCEEDED, new Metadata.Trailers());
+          }
+        }
       }
     }, timeoutMicros, TimeUnit.MICROSECONDS);
   }
 
+  private double getNextDelayMillis() {
```

Comment:
`RetryPolicy` SGTM as well. Agreed it would just have a single method. Maybe make an interface and then have `ExponentialBackoffPolicy` as an implementation ... WDYT?